reconcilers/reconcilers.go

```diff
@@ -70,9 +71,18 @@ type ParentReconciler struct {
 
 	// Reconciler is called for each reconciler request with the parent
 	// resource being reconciled. Typically, Reconciler is a Sequence of
-	// multiple SubReconcilers.
+	// multiple SubReconcilers.  It is not called on a delete request.
 	Reconciler SubReconciler
 
+	// Finalize is called on a delete request.  Typically, Finalize cleans
+	// up any state, created by previous reconciles, of which the parent is
+	// not a Kubernetes owner.
+	//
+	// Expected function signature:
+	//     func(ctx context.Context, parent client.Object) error
+	// +optional
+	Finalize interface{}
```

Comment:
I thought about this too but wonder how some subreconcilers would behave in the case of deletion.  SyncReconcilers are a natural fit (even share the same signature as what I implemented).  ChildReconcilers seem to be somewhere we do not want to have on deletion. If we were to go down that route, would we prevent ChildReconcilers some how or just rely on user restraint?

Reply:
I think we'd want finalizer support for each built-in sub reconciler. A ChildReconciler that creates a resource in a different namespace should still be able to clean up that resource upon deletion of the parent. The sub reconciler, especially when composed, will have more context about the state that needs to be cleaned up than the parent reconciler.

Reply:
Interesting, so to check I understand, given SyncReconciler and ChildReconciler as examples:
- The ChildReconciler would create a finalizer if a) itself is a namespaced scope resource, and b) it's child is either clusterscoped or in a different namespace.  On a deletion, it would then have remove finalizer, and the resource using an explicit Delete.  If it doesn't, match a) and b) it would return early before child creation
- The SyncReconciler would need to have its functionality exposed via a Finalize like method which if present is added in.  On a deletion, it would then remove the finalizer after running Finalize.

And by necessity the ParentReconciler will now allow deletion requests through to subreconcilers (unless it were to check all whether all subreconcilers need to create finalizers.

Reply:
Ive pushed a version that does part of the above:
- SyncReconciler adds a finalizer and has the new method
- ChildReconciler just ignores it and will assume same namespace tidy up via ownerreferences

The tricky bit I found was how to allow for a unique finalizer for each Sync (or any other reconciler that wanted one).  I came up with the following options:
- Hash the contents of the Finalize function
  - I don't think this works the Finalize functions shared across SubReconciler
- The name of Subreconciler via reflection
  - You could also have same named reconcilers but different packages, or inline'd SubReconcilers
- Have user pass in custom finalizers
  - feels like breaking the abstraction but allowing selection of finalizers perhaps?
- Have unique finalizer generated per reconciler via incrementing a count in the stash (I tried implemented this)
  - Had to create a UniqueFinalizer() func to handle this and make it public so other Subreconcilers could use it.

Open to thoughts on the options or suggestions of others!